back-end/helpers/get/transcript_data_grouper.spec.js

```diff
@@ -0,0 +1,417 @@
+import TranscriptDataGrouper from "./transcript_data_grouper.js";
+import testIntentFactory from "../test-helpers/testIntentFactory.js";
+import IntentDao from "../../dao/intentdao.js";
+
+// create a very fake dao
+const fakeDao = new IntentDao();
+TranscriptDataGrouper.setIntentDao(fakeDao);
+
+const mapMaker = (keysAndPairs) => {
+    const result = new Map();
+    for (const [key, value] of Object.entries(keysAndPairs)) {
+        result.set(key, value);
+    }
+    return result;
+};
+
+describe("TranscriptDataGrouper", () => {
+    let intents;
+    describe("when intents are not similar", () => {
+        // Create a variety of non-similar intents
+        const baseFakeIntents = [
+            {
+                question: "hi I am Sam",
+                children: mapMaker({
+                    "h": 1,
+                    "b": 2,
+                }),
+                total_children:3,
+                project_id: 1,
+            },
+            {
+                question: "hi I am not Bob sorry sorry sorry",
+                children: mapMaker({
+                    "h": 1,
+                    "c": 2,
+                }),
+                total_children:3,
+                project_id: 1,
+            },
+            {
+                question: "hi I Mr Bob but",
+                children: mapMaker({
+                    "h": 1,
+                    "b": 2,
+                }),
+                total_children:3,
+                project_id: 1,
+            },
+            {
+                question: "This is not similar",
+                children: mapMaker({
+                    "h": 1,
+                    "b": 2,
+                }),
+                total_children:3,
+                project_id: 1,
+            },
+        ];
+        beforeEach(() => {
+            // create a new deep copy of the array each test to prevent bleeding
+            intents = [];
+            for (const fake of baseFakeIntents) {
+                // make a deep copy of the map
+                intents.push({
+                    ...fake,
+                    children: new Map(fake.children),
+                });
+            }
+            jest.spyOn(fakeDao, "getIntent").mockImplementation((query) => {
+                return intents.filter((intent) => {
+                    return intent.question === query.question && intent.project_id === query.project_id;
+                });
+            });
+        });
+        it("should not merge 2 intents which are completely different", async() => {
+            const alreadyProcessedIntents = intents.slice(0, 1);
+            const dissimilarIntent = intents[3];
+            const grouper = new TranscriptDataGrouper(alreadyProcessedIntents, 1)
+            const result = await grouper.group(dissimilarIntent);
+            expect(result).toBe(false);
+            expect(fakeDao.getIntent).not.toHaveBeenCalled();
+        });
```

Comment:
We should add a test specifically for merging intents of different length

Reply:
What do you mean?

Reply:
There is currently a special case in the merging algorithm that automatically rejects an intent (returns 0 similarity score) if the difference in length is >= 3 iirc (or maybe just >3). We should test that this works explicitely